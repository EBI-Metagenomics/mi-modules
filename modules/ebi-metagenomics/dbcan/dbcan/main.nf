process DBCAN {
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/dbcan:4.1.4--pyhdfd78af_0' :
        'biocontainers/dbcan' }"

    input:
    tuple val(meta), path(faa), path(gff)
    tuple path(dbcan_db, stageAs: "dbcan_db"), val(db_version)

    output:
    tuple path("dbcan/", type: "dir"), val("${params.db_version}"), emit: dbcan_output
    path "versions.yml"                                      , emit: versions

    script:
    """
    run_dbcan \\
        --dia_cpu ${task.cpus} \\
        --hmm_cpu ${task.cpus} \\
        --tf_cpu ${task.cpus} \\
        --dbcan_thread ${task.cpus} \\
        --db_dir dbcan_db \\
        --out_dir dbcan \\
        --cgc_substrate \\
        --cluster ${gff} \\
        ${faa} \\
        protein

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        dbcan: "${params.dbcan_version}"
        dbcan_db: "${params.db_version}"
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def VERSION = "${params.dbcan_version}"
    def DB_VERSION = "${params.db_version}"
    """
    mkdir dbcan
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        dbcan: ${VERSION}
        dbcan_db: ${DB_VERSION}
    END_VERSIONS
    """
}
