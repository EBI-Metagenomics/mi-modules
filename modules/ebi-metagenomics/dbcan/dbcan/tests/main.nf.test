nextflow_process {

    name "Test Process DBCAN"
    script "../main.nf"
    process "DBCAN"

    tag "modules"
    tag "modules_nfcore"
    tag "dbcan"

    test("running dbCAN") {

        when {
            process {
                """
                def dbcan_version = "4.1.4"
                def db_version = "4.1.4-V13"
                input[0] = [
                    [ id:'test' ], // meta map
                    file("${moduleDir}/tests/dbcan_test_db/ecoliK12MG1655_test.faa", checkIfExists: true), // protein faa
                    file("${moduleDir}/tests/dbcan_test_db/ecoliK12MG1655_test.gff", checkIfExists: true) // protein gff
                    ]
                input[1] = [
                    file("${moduleDir}/tests/dbcan_test_db", type: 'dir', checkIfExists: true),
                    "4.1.4_V13"
                    ]
                """
            }
        }

        then {
        def stablefiles = []
            file(process.out.dbcan_output.get(0).get(0)).eachFileRecurse{ file -> if (!file.isDirectory() && !["cgc_standard.out.json", "cgc_standard.out",     "cgc.out", "cgc.gff", "PUL_blast.out", "CGC.faa"].find {file.toString().endsWith(it)}) {stablefiles.add(file)}}
        def unstablefiles = []
            file(process.out.dbcan_output.get(0).get(0)).eachFileRecurse{ file -> if (["cgc_standard.out.json", "cgc_standard.out", "cgc.out", "cgc.gff",     "PUL_blast.out", "CGC.faa"].find {file.toString().endsWith(it)}) {unstablefiles.add(file.getName().toString())}}
        assertAll(
            { assert process.success },
            {assert snapshot(process.out.versions).match("versions")},
            { assert snapshot(
                    stablefiles.sort(),
                    unstablefiles.sort(),
                    process.out.versions
                ).match() }
        )
        }
    }
}
