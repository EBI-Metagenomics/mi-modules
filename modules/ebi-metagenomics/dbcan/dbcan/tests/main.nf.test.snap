{
    "test go": {
        "content": [
            [
                "dbcan-sub.hmm.out:md5,4eeb5ef0f231141879b72b2c480987ed",
                "diamond.out:md5,3c812c9fbfd32c20bf4bf314b9ceec03",
                "hmmer.out:md5,72b7bd7720e344da6967c723e571bb80",
                "overview.txt:md5,a38606e637d917b51ec2ee10d69efed9",
                "stp.out:md5,807a434fb0ce8078e863940858c463dd",
                "substrate.out:md5,efca5ea50fef88b947ebdf89a34976a9",
                "tf-1.out:md5,d41d8cd98f00b204e9800998ecf8427e",
                "tf-2.out:md5,d41d8cd98f00b204e9800998ecf8427e",
                "tp.out:md5,e9eb66ecff04b226deeb27a8fade736a",
                "uniInput:md5,d6ccc30e774cac8b58fa4be247211d61"
            ],
            [
                "CGC.faa",
                "PUL_blast.out",
                "cgc.gff",
                "cgc.out",
                "cgc_standard.out",
                "cgc_standard.out.json"
            ],
            [
                "versions.yml:md5,783a377fa90ab564df8e08b86e53d839"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.1"
        },
        "timestamp": "2025-01-28T09:08:23.829947458"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,783a377fa90ab564df8e08b86e53d839"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.1"
        },
        "timestamp": "2025-01-28T09:23:00.019919622"
    }
}